## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [9]:
import pandas as pd

from google.colab import files
uploaded = files.upload()


Saving client_dataset.csv to client_dataset (2).csv


In [18]:
# Read data
import io
df = pd.read_csv(io.BytesIO(uploaded['client_dataset (2).csv']))

display(df)

,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,item_id,category,subcategory,unit_price,unit_cost,unit_weight,qty,line_number
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,EUD29711-63-6U,decor,wall art,1096.80,762.71,7.50,105,1
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,XDA18116-89-4A,consumables,pens,24.95,15.09,1.49,21,0
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,ABE59463-05-7E,software,project management,13.52,7.86,1.68,39,6
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,ZMM00836-65-0C,consumables,pens,36.42,24.85,1.23,29,3
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,BZX55559-12-3X,consumables,misc,195.10,108.17,46.43,20,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,UCA28532-31-0B,software,logistics,83.13,51.60,2.25,33,8
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,EBD29277-23-7U,consumables,kitchen supplies,206.59,175.46,11.70,47,0
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,EBD30093-55-9Z,consumables,bathroom supplies,65.66,57.31,4.16,475,6
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,XMZ38239-49-3D,furniture,stands,1.48,1.22,18.04,112,0


In [16]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [19]:
# Use the describe function to gather some basic statistics
df.describe()

,client_id,order_id,order_week,order_year,unit_price,unit_cost,unit_weight,qty,line_number
count,54639.000000,5.463900e+04,54639.000000,54639.000000,54639.000000,54639.000000,54639.000000,5.463900e+04,54639.000000
mean,54837.869416,5.470190e+06,11.359139,2022.993064,136.267207,99.446073,5.004116,5.702646e+02,2.979667
std,25487.438231,2.599807e+06,7.023499,0.082997,183.873135,133.164267,5.326599,1.879552e+04,2.436320
min,10033.000000,1.000886e+06,1.000000,2022.000000,0.010000,0.010000,0.000000,0.000000e+00,0.000000
25%,33593.000000,3.196372e+06,6.000000,2023.000000,20.800000,14.840000,1.450000,3.200000e+01,1.000000
50%,53305.000000,5.496966e+06,11.000000,2023.000000,68.310000,49.890000,3.240000,6.800000e+01,3.000000
75%,78498.000000,7.733869e+06,17.000000,2023.000000,173.160000,125.570000,6.890000,1.700000e+02,5.000000
max,99984.000000,9.998480e+06,52.000000,2023.000000,1396.230000,846.270000,46.430000,3.958244e+06,9.000000


In [ ]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df[['unit_price']].describe()
df.describe(include = 'all')
df.count()
df.head()

In [20]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


,count
category,
consumables,23538
furniture,11915
software,8400


In [21]:
# For the category with the most entries, which subcategory had the most entries?
df.loc [df['category'] == 'consumables' ,'subcategory'].value_counts().head(1)


,count
subcategory,
bathroom supplies,6424


In [22]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head()

,count
client_id,
33615,220
66037,211
46820,209
38378,207
24741,207


In [23]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head().index.tolist()
top_5_clients


[33615, 66037, 46820, 38378, 24741]

In [26]:
# How many total units (the qty column) did the client with the most entries order order?
total_units = df.loc[df['client_id'] == 33615, 'qty'].sum()
total_units



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [57]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['Subtotal'] = df['unit_price'] * df['qty']
df


,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,unit_cost,unit_weight,qty,line_number,Subtotal,shipping_price_x,total_price,line_cost,Profit,shipping_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,762.71,7.50,105,1,115164.00,10,126259.023250,80094.55,35080.375,404.90
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,15.09,1.49,21,0,523.95,10,981.556625,326.89,207.985,374.50
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,7.86,1.68,39,6,527.28,10,976.563900,316.54,221.665,366.60
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,24.85,1.23,29,3,1056.18,10,1297.649650,730.65,336.455,131.60
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,108.17,46.43,20,1,3902.00,10,4742.356775,2173.40,1739.525,438.83
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,...,51.60,2.25,33,8,2743.29,10,3346.316575,1712.80,1041.415,319.70
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,...,175.46,11.70,47,0,9709.73,10,11088.907775,8256.62,1464.035,440.30
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,...,57.31,4.16,475,6,31188.50,10,34476.994825,27232.25,3967.175,369.39
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,...,1.22,18.04,112,0,165.76,10,468.311050,146.64,30.045,262.90


In [58]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping_cost(order_df):
    total_weight = order_df['unit_weight'].sum()
    shipping_cost = 7 if total_weight > 50 else 10
    # Apply shipping cost per pound to total weight
    return shipping_cost * total_weight
# Group by order_id and apply the shipping cost calculation
shipping_costs = df.groupby('order_id').apply(calculate_shipping_cost).reset_index(name='shipping_price')
# Check if 'shipping_price' already exists in df
if 'shipping_price' in df.columns:
    # If it exists, drop the existing column to avoid duplicates
    df = df.drop(columns=['shipping_price'])
# Merge shipping costs back into the main DataFrame using a left merge to avoid duplicating rows
df = df.merge(shipping_costs, on='order_id', how='left') # use how='left' and on='order_id' for merging
df




<ipython-input-58-558bc433f0b9>:9: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  shipping_costs = df.groupby('order_id').apply(calculate_shipping_cost).reset_index(name='shipping_price')


,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,unit_cost,unit_weight,qty,line_number,Subtotal,shipping_price_x,total_price,line_cost,Profit,shipping_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,762.71,7.50,105,1,115164.00,10,126259.023250,80094.55,35080.375,404.90
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,15.09,1.49,21,0,523.95,10,981.556625,326.89,207.985,374.50
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,7.86,1.68,39,6,527.28,10,976.563900,316.54,221.665,366.60
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,24.85,1.23,29,3,1056.18,10,1297.649650,730.65,336.455,131.60
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,108.17,46.43,20,1,3902.00,10,4742.356775,2173.40,1739.525,438.83
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,...,51.60,2.25,33,8,2743.29,10,3346.316575,1712.80,1041.415,319.70
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,...,175.46,11.70,47,0,9709.73,10,11088.907775,8256.62,1464.035,440.30
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,...,57.31,4.16,475,6,31188.50,10,34476.994825,27232.25,3967.175,369.39
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,...,1.22,18.04,112,0,165.76,10,468.311050,146.64,30.045,262.90


In [59]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['Subtotal'] + df['shipping_price']) * (1 + 0.0925)
df



,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,unit_cost,unit_weight,qty,line_number,Subtotal,shipping_price_x,total_price,line_cost,Profit,shipping_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,762.71,7.50,105,1,115164.00,10,126259.023250,80094.55,35080.375,404.90
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,15.09,1.49,21,0,523.95,10,981.556625,326.89,207.985,374.50
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,7.86,1.68,39,6,527.28,10,976.563900,316.54,221.665,366.60
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,24.85,1.23,29,3,1056.18,10,1297.649650,730.65,336.455,131.60
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,108.17,46.43,20,1,3902.00,10,4742.356775,2173.40,1739.525,438.83
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,...,51.60,2.25,33,8,2743.29,10,3346.316575,1712.80,1041.415,319.70
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,...,175.46,11.70,47,0,9709.73,10,11088.907775,8256.62,1464.035,440.30
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,...,57.31,4.16,475,6,31188.50,10,34476.994825,27232.25,3967.175,369.39
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,...,1.22,18.04,112,0,165.76,10,468.311050,146.64,30.045,262.90


In [60]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client)
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df

,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,unit_cost,unit_weight,qty,line_number,Subtotal,shipping_price_x,total_price,line_cost,Profit,shipping_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,762.71,7.50,105,1,115164.00,10,126259.023250,80489.45,35080.375,404.90
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,15.09,1.49,21,0,523.95,10,981.556625,691.39,207.985,374.50
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,7.86,1.68,39,6,527.28,10,976.563900,673.14,221.665,366.60
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,24.85,1.23,29,3,1056.18,10,1297.649650,852.25,336.455,131.60
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,108.17,46.43,20,1,3902.00,10,4742.356775,2602.23,1739.525,438.83
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,...,51.60,2.25,33,8,2743.29,10,3346.316575,2022.50,1041.415,319.70
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,...,175.46,11.70,47,0,9709.73,10,11088.907775,8686.92,1464.035,440.30
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,...,57.31,4.16,475,6,31188.50,10,34476.994825,27591.64,3967.175,369.39
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,...,1.22,18.04,112,0,165.76,10,468.311050,399.54,30.045,262.90


In [61]:
# Create a column for the profit of each line using line cost and line price
df['Profit'] = df['total_price'] - df['line_cost']
df




,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,unit_cost,unit_weight,qty,line_number,Subtotal,shipping_price_x,total_price,line_cost,Profit,shipping_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,762.71,7.50,105,1,115164.00,10,126259.023250,80489.45,45769.573250,404.90
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,15.09,1.49,21,0,523.95,10,981.556625,691.39,290.166625,374.50
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,7.86,1.68,39,6,527.28,10,976.563900,673.14,303.423900,366.60
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,24.85,1.23,29,3,1056.18,10,1297.649650,852.25,445.399650,131.60
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,108.17,46.43,20,1,3902.00,10,4742.356775,2602.23,2140.126775,438.83
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,...,51.60,2.25,33,8,2743.29,10,3346.316575,2022.50,1323.816575,319.70
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,...,175.46,11.70,47,0,9709.73,10,11088.907775,8686.92,2401.987775,440.30
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,...,57.31,4.16,475,6,31188.50,10,34476.994825,27591.64,6885.354825,369.39
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,...,1.22,18.04,112,0,165.76,10,468.311050,399.54,68.771050,262.90


In [64]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['Subtotal'] + df['shipping_price']) * (1 + 0.0925)
df

,first,last,job,phone,email,client_id,order_id,order_date,order_week,order_year,...,unit_cost,unit_weight,qty,line_number,Subtotal,shipping_price_x,total_price,line_cost,Profit,shipping_price
0,Donald,Harding,Immunologist,793-904-7725x39308,harding.donald.7185@sullivan.com,58515,8953482,2023-04-28,17,2023,...,762.71,7.50,105,1,115164.00,10,126259.023250,80489.45,45769.573250,404.90
1,Tiffany,Myers,Music therapist,201.442.4543x942,myers.t.6537@ferguson-johnson.net,37609,8069089,2023-05-19,20,2023,...,15.09,1.49,21,0,523.95,10,981.556625,691.39,290.166625,374.50
2,Shannon,Watson,Immunologist,687.737.9424x8503,swatson8146@payne.net,57113,1902144,2023-01-29,4,2023,...,7.86,1.68,39,6,527.28,10,976.563900,673.14,303.423900,366.60
3,Nathan,Baker,Accounting technician,827-788-8123x012,bakernathan@benson.com,46554,9031802,2023-04-25,17,2023,...,24.85,1.23,29,3,1056.18,10,1297.649650,852.25,445.399650,131.60
4,Christina,Schwartz,Chiropractor,265-829-3643,christinaschwartz9252@mcconnell.com,92089,1322274,2023-05-28,21,2023,...,108.17,46.43,20,1,3902.00,10,4742.356775,2602.23,2140.126775,438.83
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
54634,Lauren,Reese,Radio producer,977.877.5272x11382,late_reese_4081@montoya-chavez.org,29043,9021716,2023-04-26,17,2023,...,51.60,2.25,33,8,2743.29,10,3346.316575,2022.50,1323.816575,319.70
54635,Derrick,Moore,Musician,358.661.5483,derrick.moore.2602@pope.info,41908,6290153,2023-02-08,6,2023,...,175.46,11.70,47,0,9709.73,10,11088.907775,8686.92,2401.987775,440.30
54636,Monica,Gutierrez,Graphic designer,294.805.9100x339,gutierrezm3195@morris.org,35176,8692622,2023-03-05,9,2023,...,57.31,4.16,475,6,31188.50,10,34476.994825,27591.64,6885.354825,369.39
54637,Wanda,Solomon,Toxicologist,(311)767-4924,solomonwanda5962@ross.org,24485,7592730,2023-01-18,3,2023,...,1.22,18.04,112,0,165.76,10,468.311050,399.54,68.771050,262.90


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [66]:
# Check your work using the totals above. Multiple lines per ID.
print(df.loc[df['order_id'] == 2742071, 'total_price'].sum())
print(df.loc[df['order_id'] == 2173913, 'total_price'].sum())
print(df.loc[df['order_id'] == 6128929, 'total_price'].sum())


140963.232025
121484.31755
757179.2699750001


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [67]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
df['client_id'].value_counts()
df['total_price']

,total_price
0,126259.023250
1,981.556625
2,976.563900
3,1297.649650
4,4742.356775
...,...
54634,3346.316575
54635,11088.907775
54636,34476.994825
54637,468.311050


In [68]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
df['summary'] = df['qty'] + df['shipping_price'] + df['total_price'] + df['Profit']

df.summary



,summary
0,172538.49650
1,1667.22325
2,1685.58780
3,1903.64930
4,7341.31355
...,...
54634,5022.83315
54635,13978.19555
54636,42206.73965
54637,911.98210


In [73]:
# column names
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'Subtotal', 'shipping_price_x', 'total_price',
       'line_cost', 'Profit', 'shipping_price', 'summary'],
      dtype='object')

In [106]:
# Format the data and rename the columns to names suitable for presentation.
column_renames = {
    'order_id': 'Order ID',
    'client_id': 'Client ID',
    'category': 'Category',
    'subcategory': 'Subcategory',
    'qty': 'Quantity',
    'unit_price': 'Unit Price',
    'unit_cost': 'Unit Cost',
    'unit_weight': 'Unit Weight',
    'Subtotal': 'Subtotal',
    'shipping_price': 'Shipping Price',
    'total_price': 'Total Price',
    'line_cost': 'Line Cost',
    'Profit': 'Profit'
    }



# Define the money columns.
money_columns = ['Subtotal', 'Shipping Price', 'Total Price', 'Line Cost', 'Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amount):
  """
  Formats a dollar amount to millions of dollars.

  Args:
    amount: The dollar amount to format.

  Returns:
    The formatted amount in millions of dollars.
  """
  # Convert amount to numeric if it's a strin
  try:
    amount = float(amount)
  except ValueError:
    # Handle cases where conversion to float fails (e.g., non-numeric strings)
    return amount  # Or return a default value like 'N/A'
  return "${:,.2f}M".format(amount / 1000000)

# Rename the columns
df = df.rename(columns=column_renames)

# Rename the columns to reflect the change in the money format.
money_columns = ['Subtotal', 'Shipping Price', 'Total Price', 'Line Cost', 'Profit']

# Apply the currency_format_millions function to only the money columns.
formatted_values = df[money_columns].applymap(currency_format_millions)
display(formatted_values)





<ipython-input-106-54a05f6cf842>:49: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  formatted_values = df[money_columns].applymap(currency_format_millions)


,Subtotal,Shipping Price,Total Price,Line Cost,Profit
0,$$$1.2e-07MMM,$0.00M,$0.13M,$0.08M,$0.05M
1,$$$0.0MMM,$0.00M,$0.00M,$0.00M,$0.00M
2,$$$0.0MMM,$0.00M,$0.00M,$0.00M,$0.00M
3,$$$0.0MMM,$0.00M,$0.00M,$0.00M,$0.00M
4,$$$0.0MMM,$0.00M,$0.00M,$0.00M,$0.00M
...,...,...,...,...,...
54634,$$$0.0MMM,$0.00M,$0.00M,$0.00M,$0.00M
54635,$$$1e-08MMM,$0.00M,$0.01M,$0.01M,$0.00M
54636,$$$3e-08MMM,$0.00M,$0.03M,$0.03M,$0.01M
54637,$$$0.0MMM,$0.00M,$0.00M,$0.00M,$0.00M


In [108]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = df.sort_values(by='Profit', ascending=False)
sorted_df



,first,last,job,phone,email,Client ID,Order ID,order_date,order_week,order_year,...,Unit Weight,Quantity,line_number,Subtotal,shipping_price_x,Total Price,Line Cost,Profit,Shipping Price,summary
21801,Julie,Anderson,Exhibition designer,5933099462,anderson_better_4704@bell.biz,78965,3833568,2023-04-29,17,2023,...,0.09,3958244,1,$$$0.00058447MMM,10,6.385385e+08,3.962601e+08,2.422784e+08,281.60,8.847754e+08
15112,Dennis,Hawkins,"Social research officer, government",001-214-888-5838x7410,hawkins.dennis.5195@lee.com,85931,7263857,2023-03-05,9,2023,...,2.21,637318,3,$$$0.00020612MMM,10,2.251880e+08,1.171649e+08,1.080231e+08,374.01,3.338488e+08
23929,Michele,Mcdonald,Emergency planning/management officer,435-931-4939x95137,mcdonaldm317@hess-jackson.com,38473,4594154,2023-03-02,9,2023,...,2.25,219645,5,$$$0.00014825MMM,10,1.619605e+08,1.247719e+08,3.718856e+07,363.80,1.993690e+08
25972,Amy,Stewart,"Psychologist, educational",979.793.4884x57205,stewartamy@francis.com,86721,6057452,2023-03-21,12,2023,...,9.22,977121,3,$$$9.81e-05MMM,10,1.071778e+08,7.137900e+07,3.579881e+07,308.80,1.439540e+08
4063,Kendra,Garrett,Pension scheme manager,808.257.2029,garrettkendra8479@jefferson.com,24741,5167984,2023-05-08,19,2023,...,1.86,122166,7,$$$5.636e-05MMM,10,6.156826e+07,3.029005e+07,3.127821e+07,210.00,9.296885e+07
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
22976,Jennifer,Dickson,Lobbyist,5778756042,dicksonj374@reyes-le.info,73797,7039520,2023-04-28,17,2023,...,0.78,107,0,$$$0.0MMM,10,1.085945e+01,8.870000e+00,1.989450e+00,7.80,1.276489e+02
6322,Jonathan,Frey,Event organiser,438.377.8088,opportunity.frey@henry.biz,26396,5450101,2023-04-11,15,2023,...,0.78,79,1,$$$0.0MMM,10,1.319740e+01,1.129000e+01,1.907400e+00,10.50,1.046048e+02
37090,Paul,Livingston,"Accountant, chartered management",762.942.2535x642,plivingston@sanchez-thompson.biz,30272,2463629,2023-02-02,5,2023,...,0.78,69,0,$$$0.0MMM,10,1.002915e+01,8.490000e+00,1.539150e+00,7.80,8.836830e+01
6946,Kimberly,May,"Surveyor, mining",001-300-314-7048x1145,manmay5844@ball.com,62559,4220711,2023-02-16,7,2023,...,0.78,20,0,$$$0.0MMM,10,8.958500e+00,8.000000e+00,9.585000e-01,7.80,3.771700e+01
